rfcs/records.md

```diff
@@ -0,0 +1,450 @@
+The field lookup data is internal and immutable; for example, it might contain a string->index dictionary to be able to quickly locate fields in internal storage.
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+The field lookup does not use `__index` or `__newindex` metamethods (or the metatable in general).
+
+Invoking methods with `:` desugars into `getmetatable(obj).method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the record has a given method as a field.
+
+> TODO: Should we use `__namecall` instead of raw MT access? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword.
+
+> TODO: The draft RFC suggests two options for the syntax; only one will be chosen in the final version
+
+Syntax A:
+
+```lua
+record Person = { name: string, age: number }
+
+-- types can be omitted and default to any
+record Point = { x, y }
+```
+
+Syntax B:
+
+```lua
+record Person(name: string, age: number)
+-- types can be omitted and default to any
+record Point(x, y)
+```
+
+This defines `Point` simultaneously as a local variable that corresponds to the shape table, and a type variable that corresponds to the record type.
+
+The resulting shape table automatically is set up to be a valid record shape, but can still be modified by adding methods to it:
+
+```lua
+function Point.__add(l, r)
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is done using call-like syntax:
+
+Syntax A:
+
+```lua
+local person = Person { name = "Bob", age = 42 }
```

Comment:
There is a potential drawback here from an external tooling perspective.
There is no way to statically know whether this is a record constructor, or a function call passing 1 table literal as the first argument.

This becomes problematic for tools such as stylua. It would attempt to put parentheses around the call, as it wouldn't be able to differentiate the two. To be able to tell the difference, you would need to now roll your own analysis/inference. If e.g. selene wanted to add lints specific to records, it could not longer do so.

A potential solution here would be to add syntax different to table literals, such as using colons instead of equals. But from a user perspective, this might be an annoying confusion...

Reply:
A suggestion I put for this in OSS is:

```suggestion
local person = Person { name: "Bob", age: 42 }
```

Along with being able to be parsed especially, this would have the added benefit of being able to force key names to be defined statically, rather than letting:

```suggestion
local person = Person { [weirdThing()] = "Bob" }
```

...pass through until it gets linted later (assuming they need to be declared statically right now, that is)

Reply:
Similarly above, forcing key names to be static can most likely simplify type checking too - no longer need to worry about ensuring `weirdThing()` has to return `keyof Person` etc. And prevention of duplicate keys being specified.

Reply:
If the key syntax is used to differentiate records from function calls, all records need to have at least one field in the call. Empty constructor `Person {}` would still be ambiguous. If default values are allowed at some point, this might be rather common

Reply:
`Person(existingTable)` would also have to be disallowed for external tools to have a chance on the special `:` syntax -- all keys must always be explicit, which is not very lua-like and would be too strict in my opinion

Reply:
I forgot to reply here but we *definitely* aren't going to overload `:` for record construction - it needs to be `=` for consistency. That said, there might still be alternative syntax options to more clearly mark the record construction as such.

Reply:
A comment made offline by zeux but adding it here for posterity: 
"We’re definitely not overloading :, But we might need some token/glyph for construction. We basically either need to use syntax for existing table construction with implied function call or need to make it obvious during parsing before we get to fields. (Or we need to go for positional, but it’s less optimal wrt readability/changes to structure/optional fields)"